### PR TITLE
Add support for strftime() builtin

### LINKED ISSUE #1337: Add a way for users to get current time
### Is your feature request related to a problem? Please describe.

`time()` printing a string to terminal is fine for some use cases but it's more useful in others to be able to `printf()` a timestamp.

Making `time()` return a string would be API breaking and also cause issues with `strncmp` (because we cannot resolve the timestamp string in kernel) so we probably don't wanna change that.

Perhaps we can make a `epoch` builtin that returns seconds since epoch? I think seconds as granularity should be enough b/c high precision timing could just use `nsecs` instead.

### Describe the solution you'd like
`epoch` builtin to return seconds since epoch.

We could insert the epoch time at system boot in a map during prog load and then use `bpf_ktime_get_ns()` helper to compute seconds since epoch.

### Describe alternative solutions or features you've considered
Making `timestamp()` return a string

### PR DESCRIPTION
**Pull request by suyuee**
This PR introduces a new builtin function `strftime`. This function makes it possible to print the current time with `printf`:

```
bpftrace -e 'i:s:1 { printf("%s\n", strftime("%H:%M:%S", nsecs)); }'
```

Although `time` can print current time on its own, printing timestamp with `printf` can produce more concise and more readable output. Resolves #1337 .
##### Checklist
- [x] Language changes are updated in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`

### CODE REVIEW

File: src/ast/semantic_analyser.cpp [hunk 1]
```diff
@@ -879,20 +879,47 @@ void SemanticAnalyser::visit(Call &call)
...
-    if (check_varargs(call, 0, 1)) {
-      if (is_final_pass()) {
-        if (call.vargs && call.vargs->size() > 0) {
+    if (check_varargs(call, 0, 2))
+    {
+      if (is_final_pass())
+      {
+        if (call.vargs == nullptr || call.vargs->size() == 0)
+        {
+          std::string fmt_default = "%H:%M:%S\n";
```
Comment (danobi) [inline src/ast/semantic_analyser.cpp:888]: This should be a constant. Maybe
```
constexpr static auto default_time_fmt = "%H...";
```
at top of file.

File: src/ast/semantic_analyser.cpp [hunk 2]
```diff
@@ -879,20 +879,47 @@ void SemanticAnalyser::visit(Call &call)
...
+          bpftrace_.time_args_.push_back(fmt_default.c_str());
+          warning("Using `time()` to print time directly is now deprecated.",
```
Comment (danobi) [inline src/ast/semantic_analyser.cpp:890]: "directly" is a bit confusing in this context. Maybe this is more clear: "Using `time()` to implicitly print time is deprecated. Print the returned value instead".

File: src/ast/semantic_analyser.cpp [hunk 3]
```diff
@@ -879,20 +879,47 @@ void SemanticAnalyser::visit(Call &call)
...
       }
     }
+    if (call.vargs == nullptr || call.vargs->size() < 2)
+    {
+      call.type = CreateNone();
+    }
+    else
+    {
+      call.type = CreateTimestamp();
+    }
```
Comment (danobi) [inline src/ast/semantic_analyser.cpp:922]: Can't you just have the `call.type = CreateTimestamp()` right after `bpftrace_.timestamp_args_.push_back(fmt.str)` and omit this entirely?

File: src/ast/semantic_analyser.cpp [hunk 4]
```diff
@@ -879,20 +879,47 @@ void SemanticAnalyser::visit(Call &call)
...
             String &fmt = static_cast<String&>(fmt_arg);
             bpftrace_.time_args_.push_back(fmt.str);
+            warning("Using `time()` to print time directly is now deprecated.",
+                    call.loc);
+          }
+        }
+        else
+        {
+          if (check_arg(call, Type::string, 0, true) &&
+              check_arg(call, Type::integer, 1, false))
```
Comment (danobi) [inline src/ast/semantic_analyser.cpp:906]: Use an `else if` instead

File: tests/semantic_analyser.cpp [hunk 5]
```diff
@@ -511,6 +510,30 @@ TEST(semantic_analyser, call_time)
...
+  test("kprobe:f { time(\"%M:%S\", 1); }", 0);
+  test("kprobe:f { time(\"%M:%S\", nsecs); }", 0);
+  test("kprobe:f { time(\"%M:%S\", \"\"); }", 10);
+  test("kprobe:f { time(1, nsecs); }", 10);
+  test("kprobe:f { time(\"%M:%S\", 1, 1); }", 1);
+  test("kprobe:f { time(1, 1, 1); }", 1);
+  test("kprobe:f { time(\"%M:%S\", \"\", 1); }", 1);
+  test("kprobe:f { $ts = time(\"%M:%S\", 1); }", 1);
+  test("kprobe:f { @ts = time(\"%M:%S\", nsecs); }", 1);
+  test("kprobe:f { @[time(\"%M:%S\", nsecs)] = 1; }", 1);
```
Comment (danobi) [inline tests/semantic_analyser.cpp:524]: I think all 3 of these need to work

File: src/ast/semantic_analyser.cpp (hunk 3 above)
Comment (ember526) [reply src/ast/semantic_analyser.cpp:922]: I was thinking that doing so (putting `call.type = CreateTimestamp()` inside the `if (is_final_pass())` block) would result in `call.type` having 'Type::none' in previous passes. I assumed it would be prudent to let it have the correct type for all passes.

Comment (danobi) [reply src/ast/semantic_analyser.cpp:922]: Ah right. Can you move the `is_final_pass()` into the inner arms of if-branches, then? It'd be nice to keep all the type-dependent logic close to each other.

**Comment by brendangregg** [comment]
What is this?

```
TEST(semantic_analyser, call_timestamp_warnings)
{
  test_for_warning("kprobe:f { time(); }", "deprecated");
  test_for_warning("kprobe:f { time(\"%M:%S\"); }", "deprecated");
  test_for_warning("kprobe:f { time(\"%M:%S\", nsecs); }", "deprecated", true);
}
```

Does this deprecate/break the old behavior? I don't see a compelling reason why we should. It's in ~10k printed copies of the BPF book that we cannot update, so no.

I also saw this (there's no reference docs, so it's not clear what you're doing):

```
          warning("Using time() to implicitly print time is deprecated. Print "
                  "the returned value instead.",
```

I called it time() because it was not strftime(). This sounds like you're turning it into strftime(). If you want strftime(), add a new thing called strftime() and leave time() alone.

**Comment by brendangregg** [comment]
Why create a new printf() format character? Why not "%s"?

See printf.cpp: we convert a lot of types to String when printed.

```
    if (arg_type == Type::ksym || arg_type == Type::usym || arg_type == Type::probe ||
        arg_type == Type::username || arg_type == Type::kstack || arg_type == Type::ustack ||
        arg_type == Type::inet)
      arg_type = Type::string; // Symbols should be printed as strings
```

**Comment by ember526** [comment]
@brendangregg  The reason we hesitated to use "%s" was that we were afraid It would send a message to user that `strftime()` returns a string and might cause some confusion. It's tempting to link "strftime()" with "string()". For example, people may end up using 'strncmp()` on the return value, which doesn't make sense.

Using "%ts" doesn't have that issue. But the problem with it is that creating a new specifier for the old friend `printf()` is not actually user-friendly either. It's hard to remember and lacks simplicity.

I think the async nature of `strftime()` makes it hard to make perfect decisions for users, since people might not always be aware that BPF programs come with many limitations.

**Comment by brendangregg** [comment]
A) If it's called strftime() then people will use it with %s. (Where did %ts come from anyway? POSIX? I'm not familiar with it.)

B) I don't really see the use case other than with printf().

C) That ship has sailed anyway: we have things like kstack() and username() that use %s, but you can't use with strcmp().

It's a tracing language, not an application language. It only needs to support enough for common case tools -- anything more complex can be deferred to BCC. The existence of BCC allows us to keep bpftrace simple.

File: src/types.cpp [hunk 6]
```diff
@@ -369,7 +369,11 @@ SizedType CreateBuffer(size_t size)
 
 SizedType CreateTimestamp()
 {
-  return SizedType(Type::timestamp, 64);
+  auto tp = SizedType(Type::timestamp, 24);
+  tp.tuple_elems.emplace_back(Type::integer, 8);
+  tp.tuple_elems.emplace_back(Type::integer, 8);
+  tp.tuple_elems.emplace_back(Type::integer, 8);
```
Comment (danobi) [inline src/types.cpp:375]: I don't think using `tuple_elems` is correct. Why is this necessary?

File: src/types.h [hunk 7]
```diff
@@ -254,7 +254,7 @@ class SizedType
   };
   bool IsTupleTy(void) const
   {
-    return type == Type::tuple;
+    return type == Type::tuple || type == Type::timestamp;
```
Comment (danobi) [inline src/types.h:257]: timestamp is not a tuple type. I don't think you can do this. Why was this necessary?

File: docs/reference_guide.md [hunk 8]
```diff
@@ -2681,6 +2682,18 @@ contains the memcopy'd value so the value at `print()` invocation time will be
 printed.  However for maps, only the handle to the map is queued up, so the
 printed map may be different than the map at `print()` invocation.
 
+# 24. `strftime()`: Formatted timestamp
+
+Syntax:
+- `strftime(const char *format, int nsecs)`
+
+This returns a formatted timestamp that is printable with `printf`. The format string should be supported by `strftime(3)`. Use format specifier "%s" when printing the return value. Note that `strftime` does not actually return a string.
```
Comment (danobi) [inline docs/reference_guide.md:2690]: The "should" makes it sound optional. "must" is more accurate. See IETF RFC 2119: http://www.faqs.org/rfcs/rfc2119.html
```suggestion
This returns a formatted timestamp that is printable with `printf`. The format string must be supported by `strftime(3)`. Use format specifier "%s" when printing the return value. Note that `strftime` does not actually return a string.
```

File: src/ast/async_event_types.h [hunk 9]
```diff
@@ -76,6 +76,22 @@ struct Time
   }
 } __attribute__((packed));
 
+struct Strftime
+{
+  uint64_t action_id;
```
Comment (danobi) [inline src/ast/async_event_types.h:81]: I think we can drop the action_id. Don't think it's necessary

File: src/ast/codegen_llvm.cpp [hunk 10]
```diff
@@ -834,6 +834,28 @@ void CodegenLLVM::visit(Call &call)
...
+
+    b_.CreateStore(b_.GetIntSameSize(time_id_, elements.at(1)),
+                   b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(1) }));
+    strftime_id_++;
+    Expression &arg = *call.vargs->at(1);
+    arg.accept(*this);
+    b_.CreateStore(expr_,
+                   b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(2) }));
+    expr_ = buf;
+    // b_.CreateLifetimeEnd(buf);
```
Comment (danobi) [inline src/ast/codegen_llvm.cpp:857]: You'll want to put this in `expr_deleter_`: https://github.com/iovisor/bpftrace/blob/master/src/ast/codegen_llvm.h#L91 .

Callers should be running `expr_deleter_`. My #1386 should make it harder to forget about calling the deleter.

File: src/ast/codegen_llvm.cpp [hunk 11]
```diff
@@ -834,6 +834,28 @@ void CodegenLLVM::visit(Call &call)
...
+    StructType *strftime_struct = b_.GetStructType(call.func + "_t",
+                                                   elements,
+                                                   true);
+
+    AllocaInst *buf = b_.CreateAllocaBPF(strftime_struct, call.func + "_t");
+    b_.CreateStore(b_.GetIntSameSize(asyncactionint(AsyncAction::strftime),
+                                     elements.at(0)),
+                   b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(0) }));
+
+    b_.CreateStore(b_.GetIntSameSize(time_id_, elements.at(1)),
```
Comment (danobi) [inline src/ast/codegen_llvm.cpp:849]: Typo?
```suggestion
    b_.CreateStore(b_.GetIntSameSize(strftime_id_, elements.at(1)),
```

File: docs/reference_guide.md [hunk 12]
```diff
@@ -2681,6 +2682,18 @@ contains the memcopy'd value so the value at `print()` invocation time will be
...
+
+Examples:
+
+```
+# bpftrace -e 'i:s:1 { printf("%s\n", strftime("%H:%M:%S", nsecs)); }'
```
Comment (danobi) [inline docs/reference_guide.md:2695]: Maybe include some of the output too

File: src/bpftrace.cpp [hunk 13]
```diff
@@ -1673,6 +1680,52 @@ std::string BPFtrace::resolve_uid(uintptr_t addr) const
...
+{
+  auto fmt = strftime_args_[strftime_id].c_str();
+  char timestr[STRING_SIZE];
+  struct tm tmp;
+  std::ifstream file("/proc/stat");
+  if (!file)
+  {
+    std::cerr << "Fail to open file /proc/stat: " << std::strerror(errno)
+              << std::endl;
+    return "";
```
Comment (danobi) [inline src/bpftrace.cpp:1694]: I'm wondering if it'd be more helpful to return `"(?)"`. Any thoughts?

Comment (ember526) [reply src/bpftrace.cpp:1694]: Yeah, totally agree. It could be confusing from users' side to see nothing printed out.

File: src/types.cpp (hunk 6 above)
Comment (ember526) [reply src/types.cpp:375]: I made it this way to avoid an error "invalid indirect read from stack off -40+8 size 16". I'll look for another solution I guess.

**Comment by ember526** [comment]
Ready for review.

File: src/ast/irbuilderbpf.cpp [hunk 14]
```diff
@@ -273,7 +273,7 @@ Value *IRBuilderBPF::CreateMapLookupElem(Value *ctx,
   CreateCondBr(condition, lookup_success_block, lookup_failure_block);
 
   SetInsertPoint(lookup_success_block);
-  if (type.IsAggregate())
+  if (type.IsAggregate() || type.IsTimestampTy())
```
Comment (fbs) [inline src/ast/irbuilderbpf.cpp:276]: I think we should either put timestamp in IsAggregate or find a new name to describe them both (and all future async events which should be handled like this).

File: src/bpftrace.cpp [hunk 15]
```diff
@@ -1673,6 +1686,52 @@ std::string BPFtrace::resolve_uid(uintptr_t addr) const
   return username;
 }
 
+std::string BPFtrace::resolve_timestamp(uint32_t strftime_id,
```
Comment (fbs) [inline src/bpftrace.cpp:1689]: Looks like this currently gives us 1s accuracy right?  I was hoping for to get closer to 100us but that should probably go  in a follow up ticket.

File: src/bpftrace.cpp [hunk 16]
```diff
@@ -1673,6 +1686,52 @@ std::string BPFtrace::resolve_uid(uintptr_t addr) const
...
+    ss >> field;
+    if (field == "btime")
+    {
+      ss >> btime;
+      if (ss.fail())
+        btime = 0;
+      break;
+    }
+  }
+  if (btime == 0)
```
Comment (fbs) [inline src/bpftrace.cpp:1716]: btime should be a "kernel constant", making it possible to read its value once and storing it (maybe before code gets loaded into the kernel even). That would remove some useless string parsing from the "hot path" .

File: src/ast/async_event_types.h [hunk 17]
```diff
@@ -76,6 +76,20 @@ struct Time
...
 
+struct Strftime
+{
+  uint64_t strftime_id;
+  uint64_t nsecs_since_boot;
+
+  std::vector<llvm::Type*> asLLVMType(ast::IRBuilderBPF& b)
+  {
+    return {
+      b.getInt64Ty(), // strftime id
```
Comment (fbs) [inline src/ast/async_event_types.h:87]:
```suggestion
      b.getInt64Ty(), // asyncid
```
nit: to keep it consistent with the othrs

File: docs/reference_guide.md [hunk 18]
```diff
@@ -2701,6 +2702,25 @@ contains the memcopy'd value so the value at `print()` invocation time will be
 printed.  However for maps, only the handle to the map is queued up, so the
 printed map may be different than the map at `print()` invocation.
 
+# 24. `strftime()`: Formatted timestamp
+
+Syntax:
+- `strftime(const char *format, int nsecs)`
+
+This returns a formatted timestamp that is printable with `printf`. The format string must be supported by `strftime(3)`. Use format specifier "%s" when printing the return value. Note that `strftime` does not actually return a string.
```
Comment (fbs) [inline docs/reference_guide.md:2710]: can you align this to the rest of the doc, 80 (or 100?) char lines

File: docs/reference_guide.md [hunk 19]
```diff
@@ -89,6 +89,7 @@ discussion to other files in /docs, the /tools/\*\_examples.txt files, or blog p
     - [21. `buf()`: Buffers](#21-buf-buffers)
     - [22. `sizeof()`: Size of type or expression](#22-sizeof-size-of-type-or-expression)
     - [23. `print()`: Print Value](#23-print-print-value)
+    - [24. `strftime()`: Formatted timestamp](#23-formatted-timestamp)
```
Comment (fbs) [inline docs/reference_guide.md:92]:
```suggestion
    - [24. `strftime()`: Formatted timestamp](#24-formatted-timestamp)
```

File: src/bpftrace.cpp (hunk 15 above)
Comment (suyuee) [reply src/bpftrace.cpp:1689]: Yes. If we want someting better, we may have to choose something more accurate than btime. `clock_gettime(CLOCK_REALTIME)` might help though I'm not so sure if that's what we want; and if we use it, it's likely we cannot allow `strftime` to accept user specified timestamp.

File: src/bpftrace.cpp (hunk 16 above)
Comment (suyuee) [reply src/bpftrace.cpp:1716]: Sure, will do.

File: src/ast/async_event_types.h (hunk 17 above)
Comment (suyuee) [reply src/ast/async_event_types.h:87]: Actually, it's not `asyncid` since `strftime()` doesn't call `CreatePerfEventOutput` directly and we don't need the `asyncid` to indicate what the data is in `perf_event_printer`. Instead, it's the index in the `bpftrace.strftime_args_` which stores the format strings of strftime.

File: docs/reference_guide.md (hunk 18 above)
Comment (suyuee) [reply docs/reference_guide.md:2710]: sure

File: src/ast/irbuilderbpf.cpp (hunk 14 above)
Comment (suyuee) [reply src/ast/irbuilderbpf.cpp:276]: I'd agree. I think `IsCompound` could be a nice name here since C++ seems to use similar term: [`std::is_compound`](https://en.cppreference.com/w/cpp/types/is_compound) and it includes array, tuple, class and so on (`Type::timestamp` seems like a class type to me).  Btw there could be some other options like `IsCollectionTy`, `IsComposite` and `IsCombined` though I'm not sure if there are as good.

Comment (suyuee) [reply src/ast/irbuilderbpf.cpp:276]: Discussed with @danobi and thought that it might be good to add a helper function instead.

File: docs/reference_guide.md [hunk 20]
```diff
@@ -2701,6 +2702,28 @@ contains the memcopy'd value so the value at `print()` invocation time will be
...
 
+## 24. `strftime()`: Formatted timestamp
+
+Syntax:
+- `strftime(const char *format, int nsecs)`
+
+This returns a formatted timestamp that is printable with `printf`. The format
+string must be supported by `strftime(3)`. Use format specifier "%s" when
+printing the return value. Note that `strftime` does not actually return a
+string.
```
Comment (fbs) [inline docs/reference_guide.md:2713]:
```suggestion
string in bpf  (kernel), the formatting happens in userspace.
```

File: src/ast/codegen_llvm.cpp [hunk 21]
```diff
@@ -26,7 +27,7 @@ namespace {
 bool shouldBeOnStackAlready(SizedType &type)
```
Comment (fbs) [inline src/ast/codegen_llvm.cpp:27]: should we move this to codegen_helper too then?

File: docs/reference_guide.md [hunk 22]
```diff
@@ -2701,6 +2702,28 @@ contains the memcopy'd value so the value at `print()` invocation time will be
 printed.  However for maps, only the handle to the map is queued up, so the
 printed map may be different than the map at `print()` invocation.
 
+## 24. `strftime()`: Formatted timestamp
+
```
Comment (fbs) [inline docs/reference_guide.md:2706]: Wonder if we should mark this experimental as we're so close to the 0.11 release What do you think @danobi

File: docs/reference_guide.md (hunk 20 above)
Comment (suyuee) [reply docs/reference_guide.md:2713]: It looks much better, thanks.

File: src/ast/codegen_llvm.cpp (hunk 21 above)
Comment (suyuee) [reply src/ast/codegen_llvm.cpp:27]: I'd agree. Sounds reasonable.

File: CHANGELOG.md [hunk 23]
```diff
@@ -50,11 +50,14 @@ and this project adheres to
   - [#1377](https://github.com/iovisor/bpftrace/pull/1377)
 - Add support for non-map print()
   - [#1381](https://github.com/iovisor/bpftrace/pull/1381)
+- Add builtin function: `strftime`
+  - [#1387](https://github.com/iovisor/bpftrace/pull/1387)
```
Comment (fbs) [inline CHANGELOG.md:54]: needs to move up due to 0.11 release
